File: frontend/src/services/conversations.ts
import { supabase } from './supabase';
import type { User } from '../types/database';

export interface ConversationParticipant {
  id: string;
  conversation_id: string;
  user_id: string;
  joined_at: string;
}

export interface Conversation {
  id: string;
  created_at: string;
  updated_at?: string;
  participants: User[];
  last_message?: any;
  last_message_at?: string;
  unread_count: number;
}

export interface ConversationListResponse {
  conversations: Conversation[];
  total: number;
}

export class ConversationService {
  /**
   * Get the API base URL
   */
  private static getApiBaseUrl(): string {
    return import.meta.env.VITE_API_BASE_URL || '';
  }

  /**
   * Get the current auth token from Supabase
   */
  private static async getAuthToken(): Promise<string> {
    const { data: { session }, error } = await supabase.auth.getSession();
    
    if (error || !session?.access_token) {
      throw new Error('Not authenticated');
    }

    const token = session.access_token;
    console.log('Using auth token:', token.substring(0, 50) + '...');
    return token;
  }

  /**
   * Create a new conversation or find existing one with another user
   */
  static async createConversation(participantUsername: string): Promise<Conversation> {
    const token = await this.getAuthToken();

    const response = await fetch(`${this.getApiBaseUrl()}/api/conversations`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify({
        participant_username: participantUsername,
      }),
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error('API Error Response:', response.status, errorText);
      try {
        const errorJson = JSON.parse(errorText);
        throw new Error(errorJson.detail || 'Failed to create conversation');
      } catch {
        throw new Error(`API Error: ${response.status} - ${errorText || 'Unknown error'}`);
      }
    }

    const responseText = await response.text();
    console.log('Create Conversation Response:', responseText);
    
    if (!responseText.trim()) {
      throw new Error('Empty response from server');
    }

    try {
      return JSON.parse(responseText);
    } catch (parseError) {
      console.error('JSON Parse Error:', parseError);
      console.error('Response Text:', responseText);
      throw new Error('Invalid JSON response from server');
    }
  }

  /**
   * Get all conversations for the current user
   */
  static async getConversations(): Promise<ConversationListResponse> {
    const token = await this.getAuthToken();

    const response = await fetch(`${this.getApiBaseUrl()}/api/conversations`, {
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error('API Error Response:', response.status, errorText);
      try {
        const errorJson = JSON.parse(errorText);
        throw new Error(errorJson.detail || 'Failed to get conversations');
      } catch {
        throw new Error(`API Error: ${response.status} - ${errorText || 'Unknown error'}`);
      }
    }

    const responseText = await response.text();
    console.log('API Response Text:', responseText);
    
    if (!responseText.trim()) {
      throw new Error('Empty response from server');
    }

    try {
      return JSON.parse(responseText);
    } catch (parseError) {
      console.error('JSON Parse Error:', parseError);
      console.error('Response Text:', responseText);
      throw new Error('Invalid JSON response from server');
    }
  }

  /**
   * Get a specific conversation
   */
  static async getConversation(conversationId: string): Promise<Conversation> {
    const token = await this.getAuthToken();

    const response = await fetch(`${this.getApiBaseUrl()}/api/conversations/${conversationId}`, {
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to get conversation');
    }

    return response.json();
  }

  /**
   * Delete a conversation
   */
  static async deleteConversation(conversationId: string): Promise<void> {
    const token = await this.getAuthToken();

    const response = await fetch(`${this.getApiBaseUrl()}/api/conversations/${conversationId}`, {
      method: 'DELETE',
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to delete conversation');
    }
  }

  /**
   * Get the other participant in a 1:1 conversation
   */
  static getOtherParticipant(conversation: Conversation, currentUserId: string): User | null {
    return conversation.participants.find(p => p.id !== currentUserId) || null;
  }

  /**
   * Format last message timestamp for display
   */
  static formatTimestamp(timestamp: string): string {
    const date = new Date(timestamp);
    const now = new Date();
    const diffInHours = (now.getTime() - date.getTime()) / (1000 * 60 * 60);

    if (diffInHours < 1) {
      const diffInMinutes = Math.floor((now.getTime() - date.getTime()) / (1000 * 60));
      return diffInMinutes <= 1 ? 'now' : `${diffInMinutes}m ago`;
    } else if (diffInHours < 24) {
      return `${Math.floor(diffInHours)}h ago`;
    } else {
      const diffInDays = Math.floor(diffInHours / 24);
      return diffInDays === 1 ? '1 day ago' : `${diffInDays} days ago`;
    }
  }
}

File: backend/app/api/routes/conversations.py
from fastapi import APIRouter, HTTPException, Depends, status

from app.dependencies import get_current_user
from app.models.conversation import ConversationCreate, ConversationResponse, ConversationListResponse
from app.models.user import User
from app.services.conversation import ConversationService
from app.utils.exceptions import ValidationError, NotFoundError, PermissionError

router = APIRouter()
conversation_service = ConversationService()


@router.post("/", response_model=ConversationResponse)
async def create_conversation(
    conversation_data: ConversationCreate,
    current_user: User = Depends(get_current_user)
):
    """Create a new conversation or find existing one with another user"""
    try:
        return await conversation_service.create_or_find_conversation(
            current_user.id, 
            conversation_data.participant_username
        )
    except ValidationError as e:
        raise HTTPException(status_code=422, detail=str(e))
    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e))
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        import traceback
        print(f"Error in create_conversation: {e}")
        print(traceback.format_exc())
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")


@router.get("/", response_model=ConversationListResponse)
async def get_conversations(
    current_user: User = Depends(get_current_user)
):
    """Get all conversations for the current user"""
    print(f"=== GET CONVERSATIONS CALLED FOR USER: {current_user.username} ===")
    try:
        result = await conversation_service.get_user_conversations(current_user.id)
        print(f"=== CONVERSATIONS SUCCESS: {len(result.conversations)} conversations ===")
        return result
    except Exception as e:
        import traceback
        print(f"=== ERROR in get_conversations: {e} ===")
        print(traceback.format_exc())
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")


@router.get("/{conversation_id}", response_model=ConversationResponse)
async def get_conversation(
    conversation_id: str,
    current_user: User = Depends(get_current_user)
):
    """Get a specific conversation"""
    try:
        return await conversation_service.get_conversation(current_user.id, conversation_id)
    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e))
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Internal server error")


@router.delete("/{conversation_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_conversation(
    conversation_id: str,
    current_user: User = Depends(get_current_user)
):
    """Delete a conversation"""
    try:
        await conversation_service.delete_conversation(conversation_id, current_user.id)
    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e))
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail="Internal server error")